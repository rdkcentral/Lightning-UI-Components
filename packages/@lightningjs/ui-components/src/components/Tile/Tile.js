/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Artwork from '../Artwork';
import Badge from '../Badge';
import Checkbox from '../Checkbox';
import Label from '../Label';
import MetadataTile from '../MetadataTile';
import ProgressBar from '../ProgressBar';
import * as styles from './Tile.styles';
import Surface from '../Surface';

export default class Tile extends Surface {
  static get __componentName() {
    return 'Tile';
  }

  static get __themeStyle() {
    return styles;
  }

  static _template() {
    return {
      ...super._template(), // Make sure surface elements are patched
      Tile: {
        Artwork: {
          type: Artwork,
          signals: {
            imageLoaded: '_imageLoaded'
          },
          mount: 0.5
        },
        Content: {
          mount: 0.5
        }
      }
    };
  }

  static get properties() {
    return [
      'artwork',
      'circle',
      'badge',
      'checkbox',
      'metadata',
      'metadataLocation',
      'persistentMetadata',
      'progressBar',
      'label',
      'src'
    ];
  }

  static get tags() {
    return [
      ...super.tags,
      'Artwork',
      'Content',
      'Tile',
      { name: 'Badge', path: 'Content.Badge' },
      { name: 'Checkbox', path: 'Content.Checkbox' },
      { name: 'Metadata', path: 'Content.Metadata' },
      { name: 'ProgressBar', path: 'Content.ProgressBar' },
      { name: 'Label', path: 'Content.Label' }
    ];
  }

  /**
   * When metadata is displayed below the Tile we need a way to tell the containing columns that it exists
   * in order for it to layout properly. This approach will not however show up in the DOM inspector
   */

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return (
      this._announce || [
        this._Metadata && this._Metadata.announce,
        this._Badge && this._Badge.announce,
        this._Label && this._Label.announce,
        this._ProgressBar && this._ProgressBar.announce
      ]
    );
  }

  _update() {
    super._update();
    this._updateTileColor();
    this._updateContent();
    this._updateArtwork();
    this._updateBadge();
    this._updateLabel();
    this._updateCheckbox();
    this._updateProgressBar();
    this._updateMetadata();
  }

  /* ------------------------------ General Tile Layout ------------------------------ */

  set h(v) {
    super.h = v;
  }

  get h() {
    return this.metadataLocation !== 'inset'
      ? super.h + ((this._Metadata && this._Metadata.h) || 0)
      : super.h;
  }

  get innerH() {
    return this._h; // Ensure that surface respects the correct height when metadata is displayed below
  }

  get _gradient() {
    if (this._isCircleLayout) return false;
    return Boolean(
      this._isInsetMetadata &&
        this._hasMetadata &&
        (this._shouldShowMetadata || this._shouldShowBadgeLabel)
    );
  }

  get _isCircleLayout() {
    return Boolean(this._itemLayout && this._itemLayout.circle);
  }

  get _foregroundDefaultWidth() {
    return parseFloat(this._w / this._h).toFixed(2) ===
      parseFloat(16 / 9).toFixed(2)
      ? this.innerW * 0.5
      : this.innerW * 0.75;
  }

  _updateTileColor() {
    this._Tile.alpha = this.style.alpha;
  }

  /* ------------------------------ Artwork ------------------------------ */
  _updateArtwork() {
    // ensure a nested artwork src takes precedence over the class's src setter,
    // but that if src is undefined in both the setter and artwork object,
    // we don't incorrectly pass "src: undefined" to the Artwork component)
    this._Artwork.patch({
      h: this._h,
      w: this._w,
      x: this._w / 2,
      y: this._h / 2,
      src: this.src,
      ...(this.artwork || {}),
      style: {
        radius: this.style?.radius,
        ...this.artwork?.style
      },
      gradient: this._gradient,
      shouldScale: this._isFocusedMode
    });
  }

  _getSrc() {
    return (this.artwork && this.artwork.src) || this._src;
  }

  _imageLoaded() {
    this._Background.alpha = 0; // Since the image is loaded the surface does not need to be shown
  }

  /* ------------------------------ Badge ------------------------------ */
  _updateBadge() {
    const badgePatch = {
      ...this.badge,
      x: this.style.paddingX,
      y: this.style.paddingY,
      alpha:
        !this._persistentMetadata || !this.badge?.title || this._isCircleLayout
          ? 0.001
          : 1
    };
    if (!this._Badge) {
      this._Content.patch({
        Badge: {
          type: Badge,
          ...badgePatch,
          signals: {
            loadedBadge: '_updateBadge'
          }
        }
      });
      return;
    }

    this.applySmooth(this._Badge, badgePatch, {
      ...badgePatch,
      ...this._badgeLabelTransitions // Badge and Label should animate in with the same values
    });
  }

  /* ------------------------------ Label  ------------------------------ */
  _updateLabel() {
    if (!this.label?.title || this._isCircleLayout) {
      if (this._Label) {
        this._Content.patch({
          Label: undefined
        });
      }
      return;
    }
    const labelPatch = {
      ...this.label,
      x: this._w - this.style.paddingX,
      y: this.style.paddingY,
      alpha:
        !this._persistentMetadata || this._isCircleLayout || !this.label.title
          ? 0.001
          : 1
    };

    if (!this._Label) {
      this._Content.patch({
        Label: {
          type: Label,
          mountX: 1,
          ...labelPatch,
          signals: {
            loadedLabel: '_updateLabel'
          }
        }
      });
      return;
    }

    this.applySmooth(this._Label, labelPatch, {
      ...labelPatch,
      x: [
        labelPatch.x,
        this._shouldShowBadgeLabel
          ? this.style.animationEntrance
          : this.style.animationExit
      ],
      ...this._badgeLabelTransitions
    });
  }
  /* ------------------------------ Badge & Label ------------------------------ */

  // Badge and Label should animate in with the same values
  get _shouldShowBadgeLabel() {
    return this._persistentMetadata || this._isFocusedMode;
  }
  get _badgeLabelTransitions() {
    return {
      y: [
        this._shouldShowBadgeLabel ? this.style.paddingY : 0,
        this.__shouldShowBadgeLabel
          ? this.style.animationEntrance
          : this.style.animationExit
      ],
      alpha: [
        this._shouldShowBadgeLabel && !this._isCircleLayout ? 1 : 0.001,
        this._shouldShowBadgeLabel
          ? this.style.animationEntrance
          : this.style.animationExit
      ]
    };
  }

  /* ------------------------------ Checkbox ------------------------------ */
  _updateCheckbox() {
    // Remove Checkbox if no longer required
    if (
      !(typeof this.checkbox?.checked === 'boolean' && this.checkbox.checked) ||
      this._isCircleLayout
    ) {
      if (this._Checkbox) {
        this._Content.patch({
          Checkbox: undefined
        });
      }
      return;
    }

    const checkboxPatch = {
      ...this.checkbox,
      x: this._w - this.style.paddingX,
      y: this._h - this.style.paddingY
    };

    if (!this._Checkbox) {
      this._Content.patch({
        Checkbox: {
          ...checkboxPatch,
          type: Checkbox,
          mount: 1
        }
      });
      return;
    }

    this.applySmooth(this._Checkbox, checkboxPatch);
  }
  /* ------------------------------ Progress Bar ------------------------------ */

  get _progressBarHeight() {
    return (
      (this._ProgressBar &&
        this._ProgressBar._getTransition('h')._targetValue +
          this.style.paddingY) ||
      0
    );
  }
  _updateProgressBar() {
    // Remove ProgressBar if no longer required
    if (
      !(
        typeof this.progressBar?.progress === 'number' &&
        this.progressBar.progress
      ) ||
      this._isCircleLayout
    ) {
      if (this._ProgressBar) {
        if (this.shouldSmooth) {
          this._ProgressBar._getTransition('alpha').once('finish', () => {
            this._removeProgressBar();
          });
          this._ProgressBar.smooth = { alpha: 0 };
        } else {
          this._removeProgressBar();
        }
      }
      return;
    }
    // --------------------------------------//
    if (this.progressBar.progress > 0) {
      const progressPatch = {
        ...this.progressBar,
        w: this._w - this.style.paddingX * 2,
        x: this._w / 2,
        y: this._h - this.style.paddingYProgress
      };

      if (!this._ProgressBar) {
        this._Content.patch({
          ProgressBar: {
            ...progressPatch,
            type: ProgressBar,
            mountX: 0.5,
            mountY: 1,
            alpha: this._hasMetadata && this.shouldSmooth ? 0.001 : 1
          }
        });

        if (this.shouldSmooth) {
          this._ProgressBar.smooth = {
            alpha: [
              1,
              { delay: this.style.animationEntrance.duration } // Wait for metadata to animate in
            ]
          };
        }
        return;
      }

      // TODO: See if we need to add animation to every property individually or can set parent
      this.applySmooth(
        this._ProgressBar,
        progressPatch,
        Object.keys(progressPatch).reduce((acc, prop) => {
          acc[prop] = [
            progressPatch[prop],
            this._isFocusedMode
              ? this.style.animationEntrance
              : this.style.animationExit
          ];
          return acc;
        }, {})
      );
    }
  }
  _removeProgressBar() {
    this._Content.patch({ ProgressBar: undefined });
    this._updateMetadata();
  }
  /* ------------------------------ Metadata  ------------------------------ */

  get _shouldShowMetadata() {
    return this._persistentMetadata || this._isFocusedMode;
  }

  get _isInsetMetadata() {
    return this._metadataLocation === 'inset';
  }
  // called in animateMetadata
  get _metadataTransitions() {
    return {
      y: [
        this._shouldShowMetadata || this._isInsetMetadata
          ? this._metadataY
          : this._h + this.style.paddingY,
        this._shouldShowMetadata
          ? this.style.animationEntrance
          : this.style.animationExit
      ],
      alpha: [
        this._shouldShowMetadata ? 1 : 0.001,
        this._shouldShowMetadata
          ? this.style.animationEntrance
          : this.style.animationExit
      ]
    };
  }

  get _hasMetadata() {
    return MetadataTile.properties.some(
      prop => this.metadata && this.metadata[prop]
    );
  }

  get _metadataY() {
    return this._isInsetMetadata
      ? this._h - this.style.paddingY - this._progressBarHeight
      : this._h + this.style.paddingY;
  }

  get _metadataPatch() {
    return {
      alpha: this._hasMetadata && this._shouldShowMetadata ? 1 : 0.001,
      mountX: 0.5,
      mountY: this._isInsetMetadata ? 1 : 0,
      marquee: this._isFocusedMode,
      w: this._w - this.style.paddingX * 2,
      x: this._w / 2,
      y: !(this._isInsetMetadata && this._isFocusedMode)
        ? this._metadataY
        : this._h + this.style.paddingY,
      ...(this.metadata || {})
    };
  }

  _updateMetadata() {
    if (
      !this._hasMetadata ||
      (this._isCircleLayout && this._metadataLocation === 'inset')
    ) {
      return;
    }

    if (!this._persistentMetadata && !this._isFocusedMode) {
      this._animateMetadata();
      return;
    }

    if (!this._Metadata) {
      // Patch in Metadata for the first time
      this._Content.patch({
        Metadata: {
          type: MetadataTile,
          signals: {
            updateComponentDimensions: '_metadataLoaded'
          },
          ...this._metadataPatch
        }
      });

      return;
    }

    this._Metadata.patch(this._metadataPatch); // Metadata should never be patched with smooth

    this._animateMetadata();
  }

  _animateMetadata() {
    if (!this._Metadata) return;
    this.applySmooth(
      this._Metadata,
      this._metadataPatch,
      this._metadataTransitions
    );

    if (!this._isFocusedMode) {
      this._resetMarqueeAnimation();
    }
  }

  // signal in updateMetadata
  _metadataLoaded() {
    this._animateMetadata();
    if (this.metadataLocation !== 'inset') this.fireAncestors('$itemChanged'); // Send event to columns/rows that the height has been updated since metadata will be displayed below the Tile
  }

  /* ------------------------------ ?????  ------------------------------ */

  _updateContent() {
    const itemContainerPatch = {
      h: this._h,
      w: this._w,
      x: this._w / 2,
      y: this._h / 2
    };

    // Make sure container animates with same values as badge, label, and metadata
    this.applySmooth(
      this._Content,
      itemContainerPatch,
      Object.keys(itemContainerPatch).reduce((acc, prop) => {
        acc[prop] = [
          itemContainerPatch[prop],
          this._isFocusedMode // this is attaching transitions to props in itemContainerPatch
            ? this.style.animationEntrance
            : this.style.animationExit
        ];
        return acc;
      }, {})
    );
  }

  _resetMarqueeAnimation() {
    const alphaTransition = this._Metadata._getTransition('alpha');
    if (alphaTransition) {
      alphaTransition.on('finish', () => {
        if (this._Metadata) {
          this._Metadata.resetMarquee();
        }
      });
    } else {
      this._Metadata.resetMarquee();
    }
  }
}
