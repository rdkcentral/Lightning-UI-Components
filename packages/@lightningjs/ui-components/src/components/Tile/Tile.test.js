/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  pathToDataURI,
  makeCreateComponent,
  fastForward
} from '@lightningjs/ui-components-test-utils';
import Tile from '.';
import Artwork from '../Artwork';
import { jest } from '@jest/globals';

const sampleImage = 'sampleImage';

const createComponent = makeCreateComponent(Tile);

describe('Tile', () => {
  let tile, testRenderer;

  beforeEach(async () => {
    [tile, testRenderer] = createComponent(
      { w: 300, h: 169, metadataLocation: 'standard' },
      {
        spyOnMethods: [
          '_update',
          '_updateArtwork',
          '_updateBadge',
          '_updateCheckbox',
          '_updateContent',
          '_updateMetadata',
          '_updateProgressBar',
          '_updateLabel'
        ]
      }
    );
    await tile.__updateSpyPromise;
    testRenderer.unfocus();
  });

  afterEach(() => {
    tile = null;
    testRenderer = null;
  });

  describe('basic', () => {
    it('renders', () => {
      const tree = testRenderer.toJSON(2);
      expect(tree).toMatchSnapshot();
    });

    it('should return the proper structure initially', () => {
      const templateObj = Tile._template();
      expect(templateObj.Tile).toMatchObject({
        Artwork: {
          type: Artwork,
          mount: 0.5
        }
      });
    });
    it('should have required properties', () => {
      expect(Tile.properties).toEqual(
        expect.arrayContaining([
          'artwork',
          'circle',
          'badge',
          'checkbox',
          'metadata',
          'metadataLocation',
          'persistentMetadata',
          'progressBar',
          'label'
        ])
      );
    });

    it('should have required tags', () => {
      expect(Tile.tags).toEqual(
        expect.arrayContaining([
          'Background',
          'Artwork',
          'Content',
          'Tile',
          { name: 'Badge', path: 'Content.Badge' },
          { name: 'Checkbox', path: 'Content.Checkbox' },
          { name: 'Metadata', path: 'Content.Metadata' },
          { name: 'ProgressBar', path: 'Content.ProgressBar' },
          { name: 'Label', path: 'Content.Label' }
        ])
      );
    });

    it('returns the proper scale', async () => {
      expect(tile.scale).toBe(1);
      testRenderer.focus();
      await tile.__updateSpyPromise;
      expect(tile.scale).toBe(1);
    });

    it('returns the proper value for default width of the foreground', async () => {
      tile.w = 300;
      tile.h = 180;
      await tile.__updateSpyPromise;
      expect(tile._foregroundDefaultWidth).toBe(tile.innerW * 0.75);
      tile.w = 300;
      tile.h = 169;
      await tile.__updateSpyPromise;
      expect(tile._foregroundDefaultWidth).toBe(tile.innerW * 0.5);
    });
  });

  describe('announce', () => {
    it('sets the announce string to the appropriate metadata announce strings', () => {
      const metadata = {
        title: 'Title',
        description: 'Description',
        logoTitle: 'Peacock'
      };
      const badge = { title: 'HD' };
      const label = { title: 'Live' };
      const progressBar = { progress: 0.5 };
      tile.patch({ metadata, badge, label, progressBar });
      tile.mode = 'focused';
      testRenderer.forceAllUpdates();
      expect(tile.announce).toEqual([
        [metadata.title, metadata.description, metadata.logoTitle],
        badge.title,
        label.title,
        `${progressBar.progress * 100}%`
      ]);
    });

    it('overrides the announce string', () => {
      const overrideString = 'Custom announce string';
      tile.announce = overrideString;
      testRenderer.forceAllUpdates();
      expect(tile.announce).toBe(overrideString);
    });
  });

  describe('artwork', () => {
    it('should accept src as getter at the root component level or nested inside artwork object', () => {
      expect(tile.src).toBeUndefined();

      const imageUrl1 = 'test.png';
      tile.src = imageUrl1;
      testRenderer.forceAllUpdates();
      expect(tile.src).toBe(imageUrl1);
      expect(tile.artwork && tile.artwork.src).toBeUndefined();
      expect(tile._Artwork.src).toBe(imageUrl1);

      const imageUrl2 = 'test.png';
      tile.artwork = { src: imageUrl2 };
      testRenderer.forceAllUpdates();
      expect(tile.src).toBe(imageUrl2);
      expect(tile.artwork.src).toBe(imageUrl2);
      expect(tile._Artwork.src).toBe(imageUrl2);

      tile.src = undefined;
      tile.artwork = { src: imageUrl2 };
      testRenderer.forceAllUpdates();
      expect(tile.src).toBe(imageUrl2);
      expect(tile.artwork.src).toBe(imageUrl2);
      expect(tile._Artwork.src).toBe(imageUrl2);
    });

    it('updates artwork in default state', async () => {
      tile._Artwork._Image.src = sampleImage;
      await tile.__updateImageSpyPromise;
      expect(tile._Artwork.constructor.name).toBe('Artwork');
      expect(tile._Artwork._Image.src).toBe('sampleImage');
      expect(tile._Artwork.mode).toBe('unfocused');
      expect(tile._Artwork.gradient).toBe(false);
      expect(tile._Artwork.w).toBe(tile._w);
      expect(tile._Artwork.h).toBe(tile._h);
      expect(tile._Artwork.x).toBe(tile._w / 2);
      expect(tile._Artwork.y).toBe(tile._h / 2);
      expect(tile._Artwork.style.radius).toBe(tile.style.radius);
    });

    it('updates artwork scale when imageScale is updated', async () => {
      tile.artwork = { src: sampleImage };
      testRenderer.focus();
      expect(tile._Artwork.style.imageScale).toBe(1); // imageScale for Artwork is 1
      tile.artwork.style = { imageScale: 2 };
      await tile.__updateArtworkSpyPromise;
      expect(tile._Artwork.style.imageScale).toBe(2);
    });

    it('should add foregroundImage if set', async () => {
      tile.artwork = {
        foregroundSrc: 'foo'
      };
      await tile.__updateArtworkSpyPromise;
      expect(tile._Artwork.foregroundSrc).toBe('foo');
    });
  });

  describe('badge', () => {
    it('should add badge if required and remove element when no longer needed', async () => {
      tile.badge = {
        title: 'test'
      };
      await tile.__updateBadgeSpyPromise;
      expect(tile._Badge).not.toBeUndefined();
      tile.itemLayout = { circle: true };
      await tile.__updateBadgeSpyPromise;
      expect(tile._Badge).toBeUndefined();
      tile.itemLayout = undefined;
      tile.badge = {
        title: 'changed'
      };
      await tile.__updateBadgeSpyPromise;
      expect(tile._Badge).not.toBeUndefined();

      tile.badge = {
        title: 'changed again'
      };
      await tile.__updateBadgeSpyPromise;
      expect(tile._Badge._transitions.title.targetValue).toBe('changed again');
      tile.shouldSmooth = false;
      tile.badge = {
        title: 'no smooth'
      };
      await tile.__updateBadgeSpyPromise;
      expect(tile._Badge.title).toBe('no smooth');
    });

    it('should not add badge if has a circle layout', async () => {
      tile.itemLayout = { circle: true };
      tile.badge = { title: 'test' };
      await tile.__updateBadgeSpyPromise;
      expect(tile._Badge).toBeUndefined();
    });
  });

  describe('circle layout', () => {
    it('should not add tag if has a circle layout', async () => {
      tile.itemLayout = { circle: true };
      tile.metadata = { tagTitle: 'test' };
      await tile.__updateTagSpyPromise;
      expect(tile._Tag).toBeUndefined();
    });

    it('returns the proper value for isCircleLayout', async () => {
      expect(tile._isCircleLayout).toBe(false);
      tile.itemLayout = {
        ratioX: 16,
        ratioY: 9,
        upCount: 5,
        circle: true
      };
      await tile.__updateSpyPromise;
      expect(tile._isCircleLayout).toBe(true);
    });
  });

  describe('label', () => {
    it('should add label if required and remove element when no longer needed', async () => {
      tile.label = {
        title: 'test'
      };
      await tile.__updateLabelSpyPromise;
      expect(tile._Label).not.toBeUndefined();
      tile.itemLayout = { circle: true };
      await tile.__updateTagSpyPromise;
      expect(tile._Tag).toBeUndefined();
      tile.itemLayout = undefined;
      tile.label = {
        title: 'changed'
      };
      await tile.__updateLabelSpyPromise;
      expect(tile._Label._transitions.title.targetValue).toBe('changed');
      tile.label = {
        title: 'changed again'
      };
      await tile.__updateLabelSpyPromise;
      expect(tile._Label._transitions.title.targetValue).toBe('changed again');
      tile.shouldSmooth = false;
      tile.label = {
        title: 'no smooth'
      };
      await tile.__updateLabelSpyPromise;
      expect(tile._Label.title).toBe('no smooth');
      tile.label = {
        title: 'with smooth'
      };
      await tile.__updateLabelSpyPromise;
      expect(tile._Label.title).toBe('with smooth');
    });
  });

  describe('Progress Bar', () => {
    it('should add a ProgressBar when progress is greater than 0 and remove it if no longer needed', async () => {
      expect(tile._ProgressBar).toBeUndefined();
      tile.progressBar = {
        progress: 0.9
      };
      await tile.__updateProgressBarSpyPromise;
      expect(tile._ProgressBar).not.toBeUndefined();
      tile.itemLayout = { w: 100, h: 200, circle: true };
      await tile.__updateProgressBarSpyPromise;
      fastForward([tile._ProgressBar]);
      testRenderer.update(); // Force redraw
      expect(tile._ProgressBar).toBeUndefined();
      tile.itemLayout = undefined;
      tile.shouldSmooth = false;
      tile.progressBar = {
        progress: 0.4
      };
      await tile.__updateProgressBarSpyPromise;
      expect(tile._ProgressBar).not.toBeUndefined();
      tile.progressBar = {
        progress: 0
      };
      await tile.__updateProgressBarSpyPromise;
      expect(tile._ProgressBar).toBeUndefined();
      tile.progressBar = {
        progress: 0.5
      };
      await tile.__updateProgressBarSpyPromise;
      expect(tile._ProgressBar).not.toBeUndefined();
      tile.shouldSmooth = true;
      tile.progressBar = {
        progress: 0
      };
      await tile.__updateProgressBarSpyPromise;
      fastForward([tile._ProgressBar]);
      testRenderer.update(); // Force redraw
      expect(tile._ProgressBar).toBeUndefined();
    });

    it('should adjust y position of metadata if inset and there is a progressbar', () => {
      tile.patch({
        progressBar: { progress: 0.5 },
        metadataLocation: 'inset',
        metadata: { title: 'test ' },
        mode: 'focused'
      });
      testRenderer.forceAllUpdates();
      expect(tile._metadataY).toBe(
        tile._ProgressBar.y - tile.style.paddingYBetweenContent
      );
    });

    it('should not patch progressBar if is in circle layout mode', async () => {
      tile.patch({
        progressBar: {
          progress: 0.5
        },
        itemLayout: {
          circle: true
        }
      });
      await tile.__updateProgressBarSpyPromise;
      expect(tile._ProgressBar).toBeUndefined();
    });
  });

  describe('checkbox', () => {
    it('should add a checkbox if required and remove the element when no longer needed', async () => {
      expect(tile._Checkbox).toBeUndefined();

      tile.checkbox = {
        checked: true
      };
      await tile.__updateCheckboxSpyPromise;
      expect(tile._Checkbox).not.toBeUndefined();

      tile.w = 100;
      await tile.__updateCheckboxSpyPromise;
      expect(tile._Checkbox._transitions.x.targetValue).toBe(70);

      tile.w = 200;
      tile.shouldSmooth = false;
      await tile.__updateCheckboxSpyPromise;
      expect(tile._Checkbox.x).toBe(170);

      tile.itemLayout = { circle: true };
      await tile.__updateCheckboxSpyPromise;
      expect(tile._Checkbox).toBeUndefined();

      tile.itemLayout = undefined;
      tile.requestUpdate(true);
      await tile.__updateCheckboxSpyPromise;
      expect(tile._Checkbox).not.toBeUndefined();

      tile.checkbox = {
        checked: false
      };
      await tile.__updateCheckboxSpyPromise;
      expect(tile._Checkbox).toBeUndefined();
    });
  });

  describe('logo', () => {
    const logo = pathToDataURI('src/assets/images/ic_lightning_white_32.png');
    it('should patch in a logo if provided', () => {
      tile.logo = logo;
      expect(tile._Icon).toBeUndefined();
    });
  });

  describe('metadata', () => {
    it('should update metadata and remove if no longer needed', async () => {
      expect(tile.metadata).toBeUndefined();

      tile.metadata = { title: 'test' };
      await tile.__updateMetadataSpyPromise;
      expect(tile.metadata).not.toBeUndefined();
      expect(tile._Metadata).not.toBeUndefined();

      tile.metadata = undefined;
      await tile.__updateMetadataSpyPromise;
      expect(tile._Metadata).toBeUndefined();
    });

    it('returns the proper value for gradient when has metadata is in focus and layout is inset', async () => {
      expect(tile._gradient).toBe(false);
      // Overwrite _hasMetadata to always be true for test
      Object.defineProperty(tile, '_hasMetadata', {
        get() {
          return true;
        }
      });
      testRenderer.focus();
      await tile.__updateSpyPromise;
      expect(tile._gradient).toBe(false);
      tile.progressBar = {
        progress: 0.5
      };
      await tile.__updateSpyPromise;
      expect(tile._gradient).toBe(true);
      tile.progressBar = {
        progress: 0
      };
      tile.metadata = { title: 'test' };

      testRenderer.unfocus();
      await tile.__updateSpyPromise;

      expect(tile._gradient).toBe(false);
      testRenderer.focus();
      await tile.__updateSpyPromise;
      expect(tile._gradient).toBe(false);
      tile.metadataLocation = 'inset';
      tile.progressBar = {
        progress: 0.7
      };
      testRenderer.focus();
      await tile.__updateSpyPromise;
      expect(tile._gradient).toBe(true);
    });
  });

  describe('marquee', () => {
    it('should reset the marquee animation when unfocused', async () => {
      const mockResetMarquee = jest.fn();
      jest.mock('../MetadataTile', () => {
        jest.fn().mockImplementationOnce(() => {
          return {
            ...jest.requireActual('../MetadataTile'),
            resetMarquee: mockResetMarquee
          };
        });
      });
      const [tile, testRenderer] = createComponent(
        {
          metadata: {
            title: 'long text'.repeat(30),
            description: 'long text'.repeat(30)
          }
        },
        {
          focused: false,
          spyOnMethods: ['_updateMetadata']
        }
      );

      await tile.__updateMetadataSpyPromise;

      expect(mockResetMarquee).not.toHaveBeenCalled();

      testRenderer.focus();
      await tile.__updateMetadataSpyPromise;

      expect(mockResetMarquee).not.toHaveBeenCalled();

      testRenderer.unfocus();

      expect(mockResetMarquee).not.toHaveBeenCalled();
    });
  });
});
