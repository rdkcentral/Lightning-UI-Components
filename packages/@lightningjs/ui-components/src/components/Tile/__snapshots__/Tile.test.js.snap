// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile renders 1`] = `
{
  "Component": {
    "Background": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Background",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Background",
      ],
      "texture": {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "Tile": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Artwork": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Image": {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 169,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Image",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Image",
              ],
              "texture": {
                "type": "ImageTexture",
              },
              "visible": true,
              "w": 300,
              "x": 0,
              "y": 0,
              "zIndex": 1,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 169,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0.5,
          "mountX": 0.5,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Artwork",
          "renderOfScreen": undefined,
          "renderToTexture": true,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Artwork",
          ],
          "type": "Artwork",
          "visible": true,
          "w": 300,
          "x": 150,
          "y": 84.5,
          "zIndex": 0,
        },
        "Content": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 169,
          "isComponent": undefined,
          "mount": 0.5,
          "mountX": 0.5,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Content",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "Content",
          ],
          "visible": true,
          "w": 300,
          "x": 150,
          "y": 84.5,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Tile",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Tile",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
